### PR TITLE
Proposal: Change MAV_CMD xml format to support UI generation

### PR DESCRIPTION
**Pull request by DonLakeFlyer**
This is not really meant to be a pull. Just a place to start discussion. Easiest way to explain this is to show how QGroundControl works with respect to editing mission items. In QGC when you are creating a mission you see something like this:

The little editor panels on the right (currently showing a Waypoint/MAV_CMD_NAV_WAYPOINT editor) are generated from a json specification. The json for MAV_CMD_NAV_WAYPOINT looks like this:

```
{
"id": 16,
"rawName": "MAV_CMD_NAV_WAYPOINT",
"friendlyName": "Waypoint",
"description": "Travel to a position in 3D space.",
"specifiesCoordinate": true,
"friendlyEdit": true,
"category": "Basic",
"param1": {
"label": "Hold:",
"units": "secs",
"default": 0,
"decimalPlaces": 0
},
"param2": {
"label": "Acceptance:",
"units": "m",
"default": 3,
"decimalPlaces": 2
},
"param3": {
"label": "PassThru:",
"units": "m",
"default": 0,
"decimalPlaces": 2
},
"param4": {
"label": "Heading:",
"units": "radians",
"default": 0.0,
"decimalPlaces": 2
}
},
```

If you look at this you'll see that it looks somewhat similar to the mavlink xml message definition with some extra attributes.

What I'd like to be able to do is to change the xml spec for mavlink MAV_CMD specification to include these extra attributes. That will then provide what I think is useful additional detail to the usage of the command parameters. But it also gives a ground station enough information to generate a user interface for the command as well.

The xsd for a param would change to this:

```
<xs:element name="param">
<xs:complexType mixed="true">
<xs:attribute ref="index" use="required"/>
<xs:attribute ref="label"/>
<xs:attribute ref="units"/>
<xs:attribute ref="decimalPlaces"/>
<xs:attribute ref="defaultValue"/>
<xs:attribute ref="minValue"/>
<xs:attribute ref="maxValue"/>
<xs:attribute ref="noLocation"/>
<xs:attribute ref="standaloneLocation"/>
</xs:complexType>
</xs:element>
```

With the additional attributes specified like this:

```
<xs:attribute name="label" type="xs:string"/> <!-- param -->
<xs:attribute name="decimalPlaces" type="xs:unsignedByte"/> <!-- param -->
<xs:attribute name="defaultValue" type="xs:float"/> <!-- param -->
<xs:attribute name="minValue" type="xs:float"/> <!-- param -->
<xs:attribute name="maxValue" type="xs:float"/> <!-- param -->
<xs:attribute name="units" type="xs:string"/> <!-- param -->
<xs:attribute name="noLocation" type="xs:boolean"/> <!-- param -->
<xs:attribute name="standaloneLocation" type="xs:boolean"/> <!-- param -->
```

The description for the new attributes is as follows:

* label - A short label for use next to a text field entry box
* decimalPlaces - The number of decimal places to show when display/editing the value
* defaultValue - Default value for the param when when a new mission item
* minValue - Minimum value for validation
* maxValue - Maximum value for validation
* units - Units for value, both for display to use and to allow for conversion "m" -> "ft"
* noLocation - true: This mission item does not specify a lat/lon location
* standaloneLocation - true: The vehicle does not fly through this item. This is so a ground station knows how to draw waypoint lines with respect to this item.

As an example, the xml message def for param 1 of MAV_CMD_NAV_WAYPOINT would change to something like this:

```
<param index="1" label="Hold" units="s" decimalPlaces="0" defaultValue="0" minValue="0">Hold time in decimal seconds. (ignored by fixed wing, time to stay at MISSION for rotary wing)</param>
```

If folks are ok with this I can go through and update all the message definitions since I already have all the data in the QGC json file. As well as create real pulls for the work.

### CODE REVIEW

**Comment by tridge** [comment]
i like this idea a lot!

**Comment by DonLakeFlyer** [comment]
This data would be used in this new QGC feature as well. Everywhere there is a value in QGC that you can change (not just parameters) you can access detailed information about it similar to help. Hoping this prevents the need for massive wiki sites. Here is an example for an App Settings (Offline Edit Cruise Speed) as well as a mission item param (where I think it will help a lot):
screen shot 2016-09-19 at 1 52 21 pm
screen shot 2016-09-19 at 1 52 58 pm
screen shot 2016-09-19 at 1 53 03 pm
screen shot 2016-09-19 at 1 53 20 pm

**Comment by DonLakeFlyer** [comment]
And the lastly I'd like to clean up the text/wording for all of the descriptions since it will all show through in QGC.

**Comment by magicrub** [comment]
This is for MAVLink 2.0 where you are extending the existing cmd, right?
You're not proposing to break v1.0 are you?

**Comment by tridge** [comment]
@magicrub it wouldn't change the wire format, just the annotation of commands. So a GCS will be able to automatically display appropriate labels in a mission editor.

**Comment by peterbarker** [comment]
A few of comments:
 - could we not have "noLocation" and just have "hasLocation"?  Yes, it will mostly be true, but may avoid double-negatives in code
 - this is relevant for my bit-rotting dataflash-units patch: could we agree on a standard to use in the "units" field?  I would like to say, "anything understood by https://en.wikipedia.org/wiki/GNU_Units (so acceleration would be "metres/(seconds*seconds)" or "metres/second/second"
 - "standAloneLocation" is also kind of a negative: "isDestination" is also not good but better IMHO.
 - "decimalPlaces" seems to indicate display properties but doesn't infer what the relevant increment is (if any).  If describing quantum properties then "0.5" would be relevant, but an increment of "1" would seem to make sense a lot of the time (for integer values)

**Comment by DonLakeFlyer** [comment]
* I was trying to avoid having to mark every entry hence the negative thing. I'm fine with noLocation and isDetination. 
* decimalPlaces is used for float clipping when shown to user to prevent them from running on. 
* I haven't added an increment but I could if folks want it. Main reason is QGC doens't support it in ui yet, so I forgot about it.

Standardizing on units makes sense since QGC uses the units naming for translation. I was planning on using the naming that QGC supports. Here are the units strings that QGC currently understands and what it can convert to (left most column is what is supported in units definitions):
```
const FactMetaData::BuiltInTranslation_s FactMetaData::_rgBuiltInTranslations[] = {
    { "centi-degrees",  "deg",  FactMetaData::_centiDegreesToDegrees,   FactMetaData::_degreesToCentiDegrees },
    { "radians",        "deg",  FactMetaData::_radiansToDegrees,        FactMetaData::_degreesToRadians },
    { "norm",           "%",    FactMetaData::_normToPercent,           FactMetaData::_percentToNorm },
};

// Translations driven by app settings
const FactMetaData::AppSettingsTranslation_s FactMetaData::_rgAppSettingsTranslations[] = {
    { "m",      "m",        false,  QGroundControlQmlGlobal::DistanceUnitsMeters,           FactMetaData::_defaultTranslator,                   FactMetaData::_defaultTranslator },
    { "meters", "meters",   false,  QGroundControlQmlGlobal::DistanceUnitsMeters,           FactMetaData::_defaultTranslator,                   FactMetaData::_defaultTranslator },
    { "m/s",    "m/s",      true,   QGroundControlQmlGlobal::SpeedUnitsMetersPerSecond,     FactMetaData::_defaultTranslator,                   FactMetaData::_defaultTranslator },
    { "m^2",    "m^2",      false,  QGroundControlQmlGlobal::AreaUnitsSquareMeters,         FactMetaData::_defaultTranslator,                   FactMetaData::_defaultTranslator },
    { "m",      "ft",       false,  QGroundControlQmlGlobal::DistanceUnitsFeet,             FactMetaData::_metersToFeet,                        FactMetaData::_feetToMeters },
    { "meters", "ft",       false,  QGroundControlQmlGlobal::DistanceUnitsFeet,             FactMetaData::_metersToFeet,                        FactMetaData::_feetToMeters },
    { "m^2",    "km^2",     false,  QGroundControlQmlGlobal::AreaUnitsSquareKilometers,     FactMetaData::_squareMetersToSquareKilometers,      FactMetaData::_squareKilometersToSquareMeters },
    { "m^2",    "ha",       false,  QGroundControlQmlGlobal::AreaUnitsHectares,             FactMetaData::_squareMetersToHectares,              FactMetaData::_hectaresToSquareMeters },
    { "m^2",    "ft^2",     false,  QGroundControlQmlGlobal::AreaUnitsSquareFeet,           FactMetaData::_squareMetersToSquareFeet,            FactMetaData::_squareFeetToSquareMeters },
    { "m^2",    "ac",       false,  QGroundControlQmlGlobal::AreaUnitsAcres,                FactMetaData::_squareMetersToAcres,                 FactMetaData::_acresToSquareMeters },
    { "m^2",    "mi^2",     false,  QGroundControlQmlGlobal::AreaUnitsSquareMiles,          FactMetaData::_squareMetersToSquareMiles,           FactMetaData::_squareMilesToSquareMeters },
    { "m/s",    "ft/s",     true,   QGroundControlQmlGlobal::SpeedUnitsFeetPerSecond,       FactMetaData::_metersToFeet,                        FactMetaData::_feetToMeters },
    { "m/s",    "mph",      true,   QGroundControlQmlGlobal::SpeedUnitsMilesPerHour,        FactMetaData::_metersPerSecondToMilesPerHour,       FactMetaData::_milesPerHourToMetersPerSecond },
    { "m/s",    "km/h",     true,   QGroundControlQmlGlobal::SpeedUnitsKilometersPerHour,   FactMetaData::_metersPerSecondToKilometersPerHour,  FactMetaData::_kilometersPerHourToMetersPerSecond },
    { "m/s",    "kn",       true,   QGroundControlQmlGlobal::SpeedUnitsKnots,               FactMetaData::_metersPerSecondToKnots,              FactMetaData::_knotsToMetersPerSecond },
};
```

**Comment by amilcarlucas** [comment]
Pull request #663 extends this to all mavlink messages

**Comment by LorenzMeier** [comment]
I'm closing this as stale - this does not mean its rejected forever, just stale. Please re-open if not.

**Comment by amilcarlucas** [comment]
Upstream pymavlink now contain the changes required to implement this

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -578,7 +578,7 @@
                <description>Commands to be executed by the MAV. They can be executed on user request, or as part of a mission script. If the action is used in a mission, the parameter mapping to the waypoint/mission message is as follows: Param 1, Param 2, Param 3, Param 4, X: Param 5, Y:Param 6, Z:Param 7. This command list is similar what ARINC 424 is for commercial aircraft: A data format how to interpret waypoint/mission data.</description>
                <entry value="16" name="MAV_CMD_NAV_WAYPOINT">
                     <description>Navigate to MISSION.</description>
-                    <param index="1">Hold time in decimal seconds. (ignored by fixed wing, time to stay at MISSION for rotary wing)</param>
+                    <param index="1" label="Hold" units="s" decimalPlaces="0" defaultValue="0" minValue="0">Hold time in decimal seconds. (ignored by fixed wing, time to stay at MISSION for rotary wing)</param>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: > seconds. (ignored

TO (remove full stop)

> seconds (ignored

**Comment by hamishwillee** [comment]
@DonLakeFlyer can you update to current master. Anyone see any reason not to merge this after that?

**Comment by DonLakeFlyer** [comment]
Isn't this all done in some other pull now?

**Comment by hamishwillee** [comment]
> Isn't this all done in some other pull now?

I have no idea. I just look at them each PR as it comes. 

AFAIK the support for parameter attributes: label units decimalPlace defaultValue minValue maxValue is now in the schema so we should be able to start rolling these values into the commands. Right @amilcarlucas ?

Actually adding them to messages is something I haven't planned on doing ...

**Comment by amilcarlucas** [comment]
This PR is just a placeholder for the discussion. AFAIK, the changes described here are not conform to the .xls schema.

This needs more commits, maybe more discussion, and in the end a big, fat squash :)

**Comment by hamishwillee** [comment]
@amilcarlucas @DonLakeFlyer 

OK, so now I'm confused - what is left to discuss/what happens next?

The param data currently in the schema are listed below:
```
<!-- definition param elements attributes (like the ones used on MAV_CMD for example) -->
<xs:attribute name="index" type="xs:unsignedByte"/>                    <!-- parameter index -->
<xs:attribute name="label" type="xs:string"/>                          <!-- parameter label (name) -->
<xs:attribute name="decimalPlaces" type="xs:unsignedByte"/>            <!-- parameter number of decimal places to be displayed in the GUI -->
<xs:attribute name="increment" type="xs:float"/>                       <!-- parameter increment -->
<xs:attribute name="minValue" type="xs:float"/>                        <!-- parameter minimum value -->
<xs:attribute name="maxValue" type="xs:float"/>                        <!-- parameter maximum value -->
```
Are we discussing whether these are "right" or not? If they are agreed, isn't the next step to roll them into the common.xml?

**Comment by amilcarlucas** [comment]
Yes, that was what I meant, this PR is just the beginning, a lot more additions are required in common.xml :)

**Comment by DonLakeFlyer** [comment]
> I suggest we merge this so you can test the QGC side of the integration?

There currently isn't any QGC integration with this yes. Going to be a large change to switch to using it. Something I want do do, but won't happen fast.

**Comment by hamishwillee** [comment]
@DonLakeFlyer The xml format is changed and live and we have some examples starting to appear. However this particular one does not appear to be complete or of real interest to you.

I am therefore closing. Reopen if you want to update head revision and push this to completion

**Pull request closed without merging**